experimental/inductor.py

```diff
@@ -0,0 +1,669 @@
+        # the original tensor if it's a materialized one, or a new example with
+        # zeros.
+        if placements is None:
+            return (
+                torch.zeros(tensor.shape, device=tensor.device)
+                if isinstance(tensor, FakeTensor) else tensor
+            )
+
+        if isinstance(tensor, FakeTensor):
+            shard = get_example_shard(tensor, mesh, default, fn=torch.zeros)
+            return DTensor.from_local(shard, device_mesh=mesh, placements=placements)
+        else:
+            # assume the input is a full real tensor, we distribute it.
+            return distribute_tensor(tensor, device_mesh=mesh, placements=placements)
+
+
+def materialize_module(
+    mod: nn.Module,
+    tensor_materializer: Callable[[FakeTensor], torch.Tensor] = _materialize_tensor,
+) -> Tuple[nn.Module, dict[torch.Tensor, torch.Tensor]]:
+    def to_real_tensor(m: nn.Module, key: str, t: T) -> Union[T, torch.Tensor, nn.Parameter]:
+        if isinstance(t, nn.Parameter):
+            return nn.Parameter(tensor_materializer(t.data, t))
+        else:
+            assert isinstance(t, torch.Tensor)
+            return tensor_materializer(t)
+
+    def materialize_this_module(m: nn.Module) -> None:
+        for key, param in m._parameters.items():
+            assert isinstance(param, nn.Parameter)
+            if True:  # if isinstance(param.data, FakeTensor):
+                dparam = to_real_tensor(m, key, param)
+                assert isinstance(dparam, nn.Parameter)
+                m.register_parameter(key, dparam)
+        for key, buffer in m._buffers.items():
+            if True:  # if isinstance(buffer, FakeTensor):
+                m._buffers[key] = to_real_tensor(m, key, buffer)
+
+    materialize_this_module(mod)
+    for name, submod in mod.named_modules():
+        materialize_this_module(submod)
+
+    return mod
+
+
+def get_example_shard(tensor, mesh, default=None, fn=torch.randn):
+    if not isinstance(tensor, Tensor):
+        return tensor
+    elif isinstance(tensor, DTensor):
+        return tensor.to_local()
+    elif hasattr(tensor, '_placements'):
+        inp_spec = DTensorSpec(mesh, tensor._placements, tensor.shape, tensor.ndim)
+        return fn(inp_spec.local_shape, device=tensor.device)
+    elif default is not None:
+        inp_spec = DTensorSpec(mesh, default, tensor.shape, tensor.ndim)
+        return fn(inp_spec.local_shape, device=tensor.device)
+    raise NotImplementedError(f'Dont know how to get a shard of tensor {tensor}')
+
+
+
+def make_inductor_fn(fn, example_inputs, default_shard_schema=None):
+    """
+    Use inductor to compile the function fn. 
+    Uses make_functional_fx to trace it passing example_inputs.
+    """
+    fnfx, attr_vals, unused_input_idx = make_functional_fx(fn)(*example_inputs)
+
+    used_args = prepare_args(attr_vals, unused_input_idx, example_inputs)
+
+    # materialize models + SPMD expansion. must be outside of fake mode
```

Comment:
I think we had talked about this and also done some initial prototyping about carrying out the SPMD expansion using FakeTensors. We should still do that IMO. Did you find any reason in your prototype for us to not pursue that?

Reply:
No fundamental reason other than it currently doesn't work :)
If we can do SPMD expansion using FakeTensors that's great.
We need to make sure that DTensor can run in eager mode with FakeTensors as well so that we can propagate sharding from parameters to optimizer states at optimizer_warmup stage.

initialize_optimizers and materialize_model may need to be slightly modified to work on FakeMode+DTensor, but  nothing fundamentally differnt (we should rename materialize_model to distribute_model since it would now take a FakeTensor() as and produce a DTensor(FakeTensor()) for parameters).